### PR TITLE
let Go merge downloaded deps instead of doing it in Python

### PR DESCRIPTION
**Pull request by slimreaper35**
JIRA: STONEBLD-685

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- n/a Docs updated (if applicable)
- n/a Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

File: cachi2/core/package_managers/gomod.py [hunk 1]
```diff
@@ -88,29 +88,42 @@ def fetch_gomod_source(request: Request) -> RequestOutput:
...
+                    }
+                )
+
+        tmp_download_cache_dir = os.path.join(tmp_dir, request.go_mod_cache_download_part)
+        log.debug(
+            "Adding dependencies from %s to %s",
+            tmp_download_cache_dir,
+            request.gomod_download_dir,
+        )
+        shutil.copytree(tmp_download_cache_dir, str(request.gomod_download_dir), dirs_exist_ok=True)
```
Comment (chmeliik) [inline cachi2/core/package_managers/gomod.py:126]: We should copy the cache only if we're not vendoring (that's what's causing most of the integration test failures)

**Comment by slimreaper35** [comment]
Please, check out the changes made in _test_gomod.py_.

**Comment by chmeliik** [comment]
Integration tests are failing:
```
  >           assert files_checksums == expected_files_checksums
  E           AssertionError: assert {} == {'gomod/pkg/mod/cache/download/rsc.io/quote/@v/v1.5.2.info': ...
```
The actual files_checksums are empty, which likely means the code either isn't copying the downloaded dependencies or is copying them to the wrong location

File: cachi2/core/package_managers/gomod.py [hunk 2]
```diff
@@ -88,29 +88,49 @@ def fetch_gomod_source(request: Request) -> RequestOutput:
...
 
-        gomod_source_path = request.source_dir / subpath
-        try:
-            gomod = _resolve_gomod(gomod_source_path, request)
-        except GoModError:
-            log.error("Failed to fetch gomod dependencies")
-            raise
+            gomod_source_path = request.source_dir / subpath
+            try:
+                gomod = _resolve_gomod(gomod_source_path, request, tmp_dir)
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py:100]: I think I found out. The TemporaryDirectory class is generic, but we don't specify the type param when subclassing it, so we lose type safety. Could you include this fix?

```diff
-class GoCacheTemporaryDirectory(tempfile.TemporaryDirectory):
+class GoCacheTemporaryDirectory(tempfile.TemporaryDirectory[str]):
```

File: cachi2/core/package_managers/gomod.py [hunk 3]
```diff
@@ -650,44 +649,6 @@ def _get_golang_pseudo_version(commit, tag=None, module_major_version=None, subp
...
-                    file_ == "list"
-                    and os.path.isfile(src_file)
-                    and os.path.exists("{}.lock".format(src_file))
-                ):
-                    _merge_files(src_file, dst_file)
-                continue
-            shutil.copy2(src_file, dst_dir)
-
-
 def _merge_files(src_file, dst_file):
```
Comment (chmeliik) [inline cachi2/core/package_managers/gomod.py:652]: Is `_merge_files` still being used somewhere or can we remove it?

**Review by chmeliik — COMMENTED** [review]
Changes LGTM

There might be one extra unused function (`_merge_files`)

It would also be great to explain a bit more about what we achieve by doing this (in the commit message). Specifically, that would be:
- the `*/@v/list` files are now merged correctly by Go
- the download cache is re-used for all modules, so we avoid downloading the same things twice

**Comment by lkolacek** [comment]
Should we name PRs with short explanation of proposed changes instead of ticket number. Since this is an upstream project, people might not be able to check our tickets in Jira. I would leave ticket number in PR description and commit message as you have it.

**Comment by brunoapimentel** [comment]
> Should we name PRs with short explanation of proposed changes instead of ticket number? Since this is an upstream project, people might not be able to check our tickets in Jira. I would leave ticket number in PR description and commit message as you have it.

+1 to this. I also think it makes our reviews easier, since we remember the context of a PR very quickly by looking at the title/summary, instead of needing to open Jira.

**Review by chmeliik — APPROVED** [review]
LGTM :tada:

**Review by ejegrova — APPROVED** [review]

**Review by lkolacek — APPROVED** [review]
Changes look great! Thank you :+1:

**Comment by chmeliik** [comment]
There's now a minor merge conflict that needs to be resolved

**Pull request merged**